### DISCUSSION TITLE
Is there a way to directly import raw pdf data?

### DISCUSSION THREAD

**Discussion by BurakUeda** [mpdf/mpdf#1678] [Q&A] [answered]
Our PDFs are retrieved from AWS S3 storage.
Currently, I am saving raw data to local storage as pdf file, use it in setSourceFile() and importPage(), and after output deleting the local file.
Is there a way to import the data directly to current PDF?

**Comment by Klap-in** [comment]
You mean the writeHTML() function of the mpdf library?

**Reply by finwe** [reply]
No, i guess they mean importing an existing PDF as a template with `setSourceFile`, but without saving an intermediary PDF file.

**Reply by BurakUeda** [reply]
Exactly. Instead of pointing to a file, setting the PDF data itself.
e.g.
`%PDF-1.4
%粤ﾏﾓ
1 0 obj
<<
/Type /Page
/MediaBox [ 0 0 595.28 841.89 ]
/Resources << /Font << /F5 2 0 R /F11 3 0 R /F16 4 0 R /F17 5 0 R /F6 6 0 R
/F10 3 0 R /F13 7 0 R >> /XObject << /Fm0 8 0 R /Fm1 9 0 R /Fm2 10 0 R >> >>
/Contents 11 0 R
/Parent 12 0 R
/Rotate 360
…
`

**Comment by jakejackson1** [comment, marked as answer]
Per the Fpdi documentation, `setSourceFile($file)`  accepts a path to the file or a stream resource or a StreamReader instance.

https://manuals.setasign.com/fpdi-manual/v2/the-fpdi-class/#method_setSourceFile

**Reply by BurakUeda** [reply]
I didn't know we can use Fpdi classes. Thank you!

**Reply by finwe** [reply]
Thanks, I've updated mPDF docs too.
